MPU6050 Task

Derivability opened the pull request:
Created bash script and kernel module. Script outputs temperature measurements from mpu6050 in Fahrenheit scale. Module creates sysfs interfaces, that provide data from mpu6050's accelerometer, gyroscope and termometer.

Commit:
MPU6050 Task: Create script file

Created empty mpu6050 script file.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Add root check

Added check to script if it's running as root

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Add device start

Added device starting by writing "0" to 107th register.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Implement temperature conversion

Implemented temperature convertion from device to Fahrenheit.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Add makefile

Added makefile for building module.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Create module source file

Created mpu6050_source.c file with module init and exit functions.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Implement probe and remove functions

Implemented mpu6050_probe and mpu6050_remove functions for driver.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Define register values

Defined needed register values and added comments.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Add read function

Added mpu6050_read_data function.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Implement accelerometer XYZ readings

Implemented sysfs interface for reading XYZ values from accelerometer.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Implement gyroscope XYZ readings

Implemented sysfs interface for reading XYZ values from gyroscope.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Implement temperature reading

Implemented temperature reading and transformation to Fahrenheit scale using byteshift operations.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

yekovalyov commented on 04-mpu6050/script/mpu6050.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+if [ "$EUID" -ne 0 ]
```
Root check is not required because i2c tools can work correctly with users from group i2c.

yekovalyov commented on 04-mpu6050/script/mpu6050.sh:
```diff
@@ -6,3 +6,17 @@ then
+	tempHexH=`i2cget -y 0 0x68 0x41`
+	tempHexL=`i2cget -y 0 0x68 0x42`
+
+	temperature=$(( -(((tempHexH << 8) + tempHexL) ^ ((1 << 0x10) - 1)) ))
```
This conversation is not correct for positive values.

Derivability replied:
Can you, please, explain why? It seems like temperature values that script outputs are pretty close to reality

yekovalyov replied:
Please consider the positive temperature values (Above 36.5 degrees in Celsius scale) Using the purposed conversation will give huge negative numbers instead of 37 degrees and above.

Derivability replied:
Got it. Thank you!

yekovalyov replied:
Is the problem solved?

yekovalyov commented on 04-mpu6050/module/mpu6050_source.c:
```diff
@@ -205,99 +224,119 @@ static ssize_t gyroscope_z_show(struct class *class, struct class_attribute *att
+{
+	mpu6050_read_data();
+	
+	sprintf(buf, "%d.%d\n", mpu6050_data.temperature_values[0], mpu6050_data.temperature_values[1]);
```
I suppose that such implementation gives wrong output to user.

yekovalyov commented on 04-mpu6050/module/mpu6050_source.c:
```diff
@@ -47,44 +53,56 @@ static struct class *attribute_class;
+
+	// Temperature transform to Fahrenheit scale
+
+	raw_temp_data = (((raw_temp_data/340) + (3653 << SHIFT)/100) * 18)/10 + (32 << SHIFT);
```
Shifting  by 8 bits and scale 0.001 is not the same.

Commit:
MPU6050 Task: Remove root check

Removed root check and replaced it with check if device already activated.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Rework temperature transform and out

Reworked temperature transformation and output to user.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
MPU6050 Task: Fix temperature conversion

Fixed temperature conversion. Improved script output to update output line instead of writing new.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

yekovalyov commented on 04-mpu6050/script/mpu6050.sh:
```diff
@@ -1,24 +1,24 @@
+	
+	if [ $sign = 1 ]
+	then
+		temperature=$(( -($temperature ^ 0xffff) ))
```
This conversation is not correct. The implemented function do conversation to 1's complement representation.
For 2's complement You can use
-(0x10000 - $temperature)
it is the same like
$temperature - 0x10000
instead of -($temperature  0xffff)

yekovalyov commented on 04-mpu6050/module/mpu6050_source.c:
```diff
@@ -68,16 +71,38 @@ static int mpu6050_read_data(void)
+        printk(KERN_INFO "mpu6050: i2c address is 0x%X\n", drv_client->addr);
+    
+    i2c_smbus_write_byte_data(drv_client, REG_CONFIG, 0);
+	i2c_smbus_write_byte_data(drv_client, REG_GYRO_CONFIG, 0);
```
It is better to use the same indentation for all lines.

yekovalyov commented:
Change label to changes requested because the issue with temperature conversation was not fixed
temperature=$(( -($temperature ^ 0xffff) ))

yekovalyov reviewed (changes requested):
Please add reply to all unresolved issues or set them as resolved if You have not doubts about the applied changes.
